stock_move_line_auto_fill/models/stock_move.py

```diff
@@ -9,21 +9,30 @@
 class StockMove(models.Model):
     _inherit = "stock.move"
 
+    def _get_auto_fill_flag(self):
+        return self.picking_id.auto_fill_operation
+
+    def _get_avoid_lot_assignment_flag(self):
+        return self.picking_id.picking_type_id.avoid_lot_assignment
+
+    def _check_auto_fill_conditions(self, move_line_vals):
+        auto_fill = self._get_auto_fill_flag()
+        avoid_lot_assignment = self._get_avoid_lot_assignment_flag()
+        lot_id = move_line_vals.get("lot_id")
+        return auto_fill and not (avoid_lot_assignment and lot_id)
+
+    def _update_qty_done(self, move_line_vals):
+        if self._check_auto_fill_conditions(move_line_vals):
+            move_line_vals["qty_done"] = move_line_vals.get("reserved_uom_qty", 0.0)
+
     def _prepare_move_line_vals(self, quantity=None, reserved_quant=None):
         """
         Auto-assign as done the quantity proposed for the lots.
         Keep this method to avoid extra write after picking _action_assign
         """
-        self.ensure_one()
-        res = super()._prepare_move_line_vals(quantity, reserved_quant)
-        if not self.picking_id.auto_fill_operation:
-            return res
-        elif self.picking_id.picking_type_id.avoid_lot_assignment and res.get("lot_id"):
-            return res
-        if self.quantity_done != self.product_uom_qty:
-            # Not assign qty_done for extra moves in over processed quantities
-            res.update({"qty_done": res.get("reserved_uom_qty", 0.0)})
-        return res
+        move_line_vals = super()._prepare_move_line_vals(quantity, reserved_quant)
+        self._update_qty_done(move_line_vals)
```

Comment:
To avoid performance regression.
```suggestion
        if not self._get_auto_fill_flag()
            return move_line_vals
        if self._get_avoid_lot_assignment_flag() and move_line_vals.get("lot_id"):
            return move_line_vals
        if self.quantity_done != self.product_uom_qty:
            # Not assign qty_done for extra moves in over processed quantities
            move_line_vals.update({"qty_done": move_line_vals.get("reserved_uom_qty", 0.0)})
```